### PR TITLE
Add document:upsert

### PR DESCRIPTION
**Pull request by y-abs**
## What does this PR do?

Add document:upsert API action https://github.com/kuzzleio/sdk-dart/issues/55

### CODE REVIEW

File: lib/src/controllers/document.dart [hunk 1]
```diff
@@ -445,7 +447,8 @@ class DocumentController extends KuzzleController {
       index: index,
       collection: collection,
       uid: id,
-      body: body,
+      body: changes,
```
Comment (Aschen) [inline lib/src/controllers/document.dart]: The body contains both `changes` and `defaults`

File: lib/src/kuzzle/request.dart [hunk 2]
```diff
@@ -200,7 +203,9 @@ class KuzzleRequest {
     if (source != null) {
       map['source'] = source;
     }
-
+    if (defaults != null) {
+      map['defaults'] = defaults;
```
Comment (Aschen) [inline lib/src/kuzzle/request.dart]: I'm not sure to understand what this is suppose to do but `defaults` is part of the body and it looks like you put it in the arguments

**Review by Aschen — CHANGES_REQUESTED** [review]

File: doc/2/controllers/document/upsert/index.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+---
+code: true
+type: page
+title: upsert
```
Comment (Aschen) [inline doc/2/controllers/document/upsert/index.md:4]: Missing page description

File: doc/2/controllers/document/upsert/index.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+## Returns
+
+A `Map<String, dynamic>` with the following properties:
+
+| Property     | Type                                         | Description                      |
+|------------- |--------------------------------------------- |--------------------------------- |
+| `_source`    | <pre>Map<String, dynamic></pre> | Updated document (if `source` option set to true)  |
+| `_id`        | <pre>String</pre>                            | ID of the udated document                   |
+| `_version`   | <pre>int</pre>                           | Version of the document in the persistent data storage |
+| `created`     | <pre>bool</pre>
```
Comment (Aschen) [inline doc/2/controllers/document/upsert/index.md]: Missing description for `created`

File: lib/src/controllers/document.dart [hunk 5]
```diff
@@ -414,6 +414,51 @@ class DocumentController extends KuzzleController {
...
   }
 
+  /// ####Applies a partial update to an existing document.
+  /// If the document doesn't already exist, a new document is created.
+  ///
+  /// **[index]**: index name \\\
+  /// **[collection]**: collection name \\\
+  /// **[id]**: unique identifier of the document to update \\\
+  /// **[changes]**: Partial changes to apply to the document
+  /// and Fields to add to the document if it gets created (optional)
```
Comment (Aschen) [inline lib/src/controllers/document.dart]:
```suggestion
```

**Review by Aschen — CHANGES_REQUESTED** [review]

**Review by xbill82 — APPROVED** [review]

**Review by Aschen — APPROVED** [review]

**Pull request merged**